### PR TITLE
Batch query eval

### PR DESCRIPTION
**Pull request by MJHamar**
good to go from my side

USAGE:
- constructor:
  - list of model names from huggingface + "bm25" if you want to use it.
  - list of weights for each model to use in the weighted sum (see `QueryEval.score()`)
- load_dataset: given a DataFrame (following the format of `.dataset.load_corpus`), compute embeddings for each model and index the corpus for the BM25 model if specified.
- save_to_cache: save the embeddings for the models and the indexed corpus for the BM25 model
- load_from_cache: 
- score: given a list of queries $(q_1,q_2,\dots,q_n)$ and a list of documents $(d_1,d_2,\dots,d_n)$, compute 

image

In words, the score function takes a bunch of query-document pairs, and for each such pair, it computes a list of similarity scores (in the range $[0,1]$) and takes their weighted average.

NOTE:
- The `score` function *approximates the likelihood* that a given query refers to a given document. It does not "rank" the documents in any traditional way.
- For dense models, we don't want to compute similarity scores for all documents in the corpus because it is expensive. Instead, we take the *cosine similarity* (scaled to $[0,1]$) and regard it as it if were a probability.
- We cannot do the same with the BM25 score, because its output domain is unbounded. So, we *have to* score all documents in the corpus, compute softmax and take the value computed for the given document.
- Currently, we control the weights and the list of models that we use in the constructor. In the future, we might want to consider fitting the weights to the training set (because, since cosine similarity $\neq$ probability, the scores might not be on the same scale still. A good way to overcome this is by fitting the scores on actual query-document pairs)

### CODE REVIEW

File: .vscode/settings.json
Comment (m-krastev) [inline .vscode/settings.json:1]: smh

File: tests/test_query_eval.py
Comment (m-krastev) [inline tests/test_query_eval.py:1]: The internship changed you, friend.

Comment (MJHamar) [reply tests/test_query_eval.py:1]: Or maybe i just got old

File: inpars/query_eval.py [hunk 1]
```diff
@@ -38,126 +41,198 @@ def _get_embeddings(self, model_name, documents: List[str], batch_size: int) ->
...
+            doc_indices = [doc_indices]
+        if len(queries) != len(doc_indices):
+            raise ValueError("Number of queries and document indices should be the same.")
+
+        # find the embeddings corresponding to the doc_indices in the dataframes
+        doc_embeddings = OrderedDict({
+            model_name: torch.tensor(
+                self.doc_embeddings[model_name].loc[
+                    self.doc_embeddings[model_name]['doc_id'].isin(doc_indices), 'embedding'
+                ].values.tolist(),
```
Comment (m-krastev) [inline inpars/query_eval.py]: Not sure what the speed is, but I don't recommend using Pandas key checks unless we don't use this query evaluation too frequently. Remember how this slowed down the FACT MCTS node value calculation.

File: inpars/query_eval.py [hunk 2]
```diff
@@ -26,8 +29,8 @@ def __init__(
             else:
                 self.tokenizers[model] = AutoTokenizer.from_pretrained(model)
                 self.models[model] = AutoModel.from_pretrained(model).to(self.device)
-        self.weights = torch.tensor(weights, device=self.device)
-        self.doc_embeddings = None
+        self.weights = OrderedDict({model: weight for model, weight in zip(models, weights)})
+        self.doc_embeddings = OrderedDict()
```
Comment (m-krastev) [inline inpars/query_eval.py]: Not sure what difference might an OrderedDict make here, considering that [normal dicts are almost the same](https://stackoverflow.com/a/58676384).

**Review by m-krastev — CHANGES_REQUESTED** [review]

File: inpars/query_eval.py (hunk 1 above)
Comment (MJHamar) [reply inpars/query_eval.py]: fair enough, i suck at pandas sorry.

File: inpars/query_eval.py (hunk 2 above)
Comment (MJHamar) [reply inpars/query_eval.py]: just making 100% sure nothing gets messed up. Ideally I'd like to get rid of dictionaries completely.

File: inpars/query_eval.py (hunk 1 above)
Comment (m-krastev) [reply inpars/query_eval.py]: I suggest having some dict mapping (or numpy array) for doc_id (or index of the doc_id) to index. 
If you use that, then you can always use torch/numpy indexing.

**Comment by MJHamar** [comment]
Will fix in the morning

**Comment by MJHamar** [comment]
got rid of dataframes and simplified to a single dense model

**Review by m-krastev — APPROVED** [review]
Awesome work. Thanks!
